core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceMarshallerImpl.java

```diff
@@ -249,19 +270,43 @@ private Object wrapUserObject(Object o) {
    }
 
    private int sizeEstimate(Object o, boolean persistenceClass) {
-      // hardcoded for now in order to avoid requiring a dependency on com.google.protobuf.CodedOutputStream
-      // Dynamic estimates will be provided in future protostream version
-      int wrapperEstimate = 40;
       if (persistenceClass) {
          if (o instanceof UserBytes) {
             byte[] user = ((UserBytes) o).bytes;
-            return wrapperEstimate + user.length;
+            return calculateWrappedUserObjectSize(user.length);
          }
          // Return the CodedOutputStream.DEFAULT_BUFFER_SIZE as this is equivalent to passing no estimate
+         // Dynamic estimates will be provided in a future protostream version IPROTO-89
          return 4096;
       }
       int userBytesEstimate = userMarshaller.getBufferSizePredictor(o.getClass()).nextSize(o);
-      return wrapperEstimate + userBytesEstimate;
+      return calculateWrappedUserObjectSize(userBytesEstimate);
+   }
+
+   private int calculateWrappedUserObjectSize(int objectBytes) {
+      int typeId = ProtoStreamTypeIds.CORE_START + 11;
+      int typeIdSize = tagSize(19, 1) + computeUInt32SizeNoTag(typeId);
+      int userBytesFieldSize = tagSize(1, 2) + computeUInt32SizeNoTag(objectBytes) + objectBytes;
+      int wrappedMessageSize = tagSize(17, 2) + computeUInt32SizeNoTag(objectBytes);
+
+      return typeIdSize + userBytesFieldSize + wrappedMessageSize;
+   }
+
+   private int tagSize(int fieldNumber, int wireType) {
+      return computeUInt32SizeNoTag(fieldNumber << 3 | wireType);
+   }
+
+   // Protobuf logic included to avoid requiring a dependency on com.google.protobuf.CodedOutputStream
+   private int computeUInt32SizeNoTag(int value) {
+      if ((value & -128) == 0) {
+         return 1;
+      } else if ((value & -16384) == 0) {
+         return 2;
+      } else if ((value & -2097152) == 0) {
+         return 3;
+      } else {
```

Comment:
Should this throw an `InvalidArgumentException`? I thought ids can only be 65K or less?